### PR TITLE
Removed unused E0406 and unused functions

### LINKED ISSUE #34301: Reassign E0406 to be thrown when undeclared associated type appears in function signature.
Currently, when an undeclared associated type is used in a function signature within a trait body, like:
```rust
trait Foo {
    type Bar;
    //error: Baz is used but not declared
    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
}
```
it throws [E0220](https://doc.rust-lang.org/error-index.html#E0220), which isn't really helpful in this situation. Also, I found [E0406](https://github.com/rust-lang/rust/blob/7d2f75a953b5645d3a336b2978b48b60d310bf54/src/librustc_resolve/diagnostics.rs#L1108) to be the more appropriate error here. This error is thrown for some feature which hasn't been implemented yet( and probably won't be), so it can be repurposed. See [#34230](https://github.com/rust-lang/rust/pull/34230#issuecomment-226182208) for more.
   We could modify the error message for `E0220` to include the above erroneous code, but I think it would be better to throw `E0406` instead of `E0220` in this situation.

/cc @GuillaumeGomez

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #34301.

It removes the unused E0406 and also unused functions.

cc @eddyb

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -1631,6 +1635,12 @@ impl<'a> Resolver<'a> {
                                         visit::walk_trait_item(this, trait_item)
                                     }
                                 }
+                                TraitItemKind::Type(..) => {
+                                    this.with_type_parameter_rib(NoTypeParameters, |this| {
+                                        visit::walk_trait_item(this, trait_item)
+                                    });
+                                    types.push(format!("{}", trait_item.ident));
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: You don't need to convert them to strings, comparing `Name`s is `O(1)`.

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -1639,14 +1649,38 @@ impl<'a> Resolver<'a> {
                                     this.with_type_parameter_rib(type_parameters, |this| {
                                         visit::walk_trait_item(this, trait_item)
                                     });
-                                }
-                                TraitItemKind::Type(..) => {
-                                    this.with_type_parameter_rib(NoTypeParameters, |this| {
-                                        visit::walk_trait_item(this, trait_item)
-                                    });
+                                    methods.push(sig.clone());
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: The diff is confusing because of how the match arms were moved around.

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -1639,14 +1649,38 @@ impl<'a> Resolver<'a> {
...
+                                for arg in method.decl.inputs.iter() {
+                                    if let Some(paths) = arg.ty.get_path() {
+                                        for path in paths {
+                                            if path.segments.len() == 2 &&
+                                               &format!("{}",
+                                                        path.segments[0].identifier) == "Self" {
+                                                let name = format!("{}",
+                                                                   path.segments[1].identifier);
+                                                let mut found = false;
+                                                for ty in types.iter() {
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: This is `types.iter().any(|ty| ty == name)`

**Comment by eddyb** [comment]
This breaks associated types that come from a different trait which `Self` also implements.
And it only looks at argument types, without even recursing into them.

I don't know what `E0406` is, but I don't see a point in producing an error earlier in the pipeline.
Resolve doesn't resolve associated types, that's a job for typeck, which knows best.

**Comment by GuillaumeGomez** [comment]
> This breaks associated types that come from a different trait which Self also implements.
> And it only looks at argument types, without even recursing into them.

It does recurse, I added the `get_path` method. Take a look into it.

**Comment by eddyb** [comment]
@GuillaumeGomez Oh hah, I missed that because of singular/plural.
As discussed on IRC, I believe the path forward is to remove the leftover `{visit,resolve}_generics`, E0406 and update E0220's description to be more general.

**Comment by cynicaldevil** [comment]
I can take up the task of updating `E0220`. Should I move forward with it or wait until this PR is merged?

**Comment by GuillaumeGomez** [comment]
No, you can do it as soon as you want. The PR will just remove unused mechanisms.

File: src/librustc_resolve/lib.rs [hunk 4]
```diff
@@ -1640,11 +1639,6 @@ impl<'a> Resolver<'a> {
                                         visit::walk_trait_item(this, trait_item)
                                     });
                                 }
-                                TraitItemKind::Type(..) => {
-                                    this.with_type_parameter_rib(NoTypeParameters, |this| {
-                                        visit::walk_trait_item(this, trait_item)
-                                    });
-                                }
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: I think this change (moving the match arms) is unnecessary.

**Pull request merged**